### PR TITLE
Implement distribution-specific build

### PR DESCRIPTION
**Pull request by Scrumplex**
To allow distribution-specific builds, mainly for third-party packages (#2), this PR allows setting some options in compile time.

**Every build can now define a `distribution`:**
- It is set to `custom` by default and `official` when running `make release`
- Third party packages can set that to an identifier to differentiate their builds from upstream
- This `distribution` tag is shown in the logger, as well as in the UI (See screenshot below)

**`updateURL` and `publicKeyString` need to be defined:**
- `updateURL` and `publicKeyString` are unset by default
  - If they are unset, the updater gets disabled
- `make release` sets those to the values we are used to
- Third party packages can either use their own values, or leave them out
  - Again, leaving them out disables the updater

**`version` needs to be defined:**
- `version` is set to `unknown` by default
- `make release` sets it to the output of `git describe --tags`
- `make dev` keeps it unset

# TODOs
- [x] Remove `updater_noop.go`

# Screenshots

Screenshot of UI version

### CODE REVIEW

**Comment by lawl** [comment]
Thanks. This looks very reasonable from a first glance (I'll have to test it still and check it in detail). One thing i noticed is that my public key is currently also hardcoded in https://github.com/lawl/NoiseTorch/blob/master/scripts/signer.go#L60

I think it would make sense to also not hardcode it there since it's now configured at build time.

Additionally, since you now inject the version build with `-X`, would it make sense to remove the go generate for the version also and inject the entire version string that way?

**Comment by Scrumplex** [comment]
> I think it would make sense to also not hardcode it there since it's now configured at build time.

Sounds good, will look into it!

> Additionally, since you now inject the version build with `-X`, would it make sense to remove the go generate for the version also and inject the entire version string that way?

Sure!

**Comment by Scrumplex** [comment]
I am not sure if I tested this right:
1. Download current signature as I can't generate one obviously: `curl --output-dir bin/ -O https://noisetorch.epicgamer.org/NoiseTorch_x64.tgz.sig`
2. Run signer in verify mode: `go run scripts/signer.go -k "3mL+rBi4yBZ1wGimQ/oSQCjxELzgTh+673H4JdzQBOk="`
I get `Verified false`

Don't know what I am doing wrong.

**Comment by Scrumplex** [comment]
Right now `make dev` does not set a version number, should I still add that?

**Comment by lawl** [comment]
I think that would be good, yeah.

File: main.go [hunk 1]
```diff
@@ -41,6 +40,11 @@ type device struct {
 }
 
 const appName = "NoiseTorch"
+var version = "unknown"
+var distribution = "custom"  // will be changed by build
+
+var updateURL = ""  // ditto
+var publicKeyString = ""  // ditto
```
Comment (lawl) [inline main.go:47]: Please run `gofmt` to fix the formatting. And the "will be changed by build" comment should probably be moved up one line.

File: update.go [hunk 2]
```diff
@@ -41,6 +43,9 @@ func updateCheck(ctx *ntcontext) {
 }
```
Comment (lawl) [inline update.go:44]: (Github won't let me comment on a line outside the diff...)

https://github.com/lawl/NoiseTorch/blob/c6de9619d668f72f284b6b6a257f0d6eac9f88d6/update.go#L103

This is not true anymore. A `panic()` could be recovered. I think if the public key decoding fails we should just err on the side of caution and `os.Exit(1)` to ensure no further update code is ran under any circumstance.

**Review by lawl — CHANGES_REQUESTED** [review]
Looked at it a bit more in depth 2 minor things, github won't let me push to your branch, or i'd have just fixed them myself.

Once those are addressed i'll move to testing.

File: update.go (hunk 2 above)
Comment (lawl) [reply update.go:44]: Should probably log it too.

**Comment by Scrumplex** [comment]
What about the following behaviour: Make the public key optional. That way we don't need to fail if the distributor decides to not sign their builds at all

**Comment by lawl** [comment]
No. The public key is already optional. If you don't specify one the updater isn't triggered and e.g. the distributions package manager handles updates and signatures. I don't want a codepath where the updater is used but the signature isn't verified.

**Comment by lawl** [comment]
Merged, thank you. Is there any organisation or project you'd like me to credit beyond yourself in the patch notes once this rolls out?

**Comment by Scrumplex** [comment]
> Is there any organisation or project you'd like me to credit beyond yourself in the patch notes once this rolls out?

Thanks for the offer!

I made this PR to solve the heated situation around the AUR package. 
I am not affiliated with the people behind the AUR package(s) nor am I an Arch Linux Developer or Trusted User (yet :D).

I am just someone from the Arch Linux community.
I do maintain [a few PKGBUILDs on the AUR](https://aur.archlinux.org/packages/?SeB=m&K=Scrumplex) myself and always (try to) contribute to upstream.
This is actually one of the core principles of Arch: https://wiki.archlinux.org/title/Arch_Linux#Simplicity

So something along the lines of:

> - ...
> - Support distribution-specific builds. @Scrumplex, from the Arch Linux community
> - ...

should be okay I think.

**Pull request merged**